tap_harvest/streams/abstracts.py

```diff
@@ -0,0 +1,281 @@
+        """Interacts with api client interaction and pagination."""
+        page = 1
+        while page:
+            self.params["page"] = page
+            response = self.client.get(
+                self.url_endpoint, self.params, self.headers, self.path
+            )
+            raw_records = response.get(self.data_key, [])
+
+            page = response.get(self.next_page_key)
+            yield from raw_records
+
+    def write_schema(self):
+        """Write a schema message."""
+        try:
+            write_schema(self.tap_stream_id, self.schema, self.key_properties)
+        except OSError as err:
+            LOGGER.error(f"OS Error while writing schema for: {self.tap_stream_id}")
+            raise err
+
+    def update_params(self, **kwargs):
+        """Update params for the stream."""
+        if self.support_filter:
+            self.params.update(kwargs)
+
+    def add_object_to_id(self, record: Dict) -> Dict:
+        """Add object_to_id to the stream."""
+        for key in self.object_to_id:
+            if record[key] is not None:
+                record[key + "_id"] = record[key]["id"]
+            else:
+                record[key + "_id"] = None
+
+        return record
+
+    def modify_object(self, record: Dict, parent_record: Dict = None) -> Dict:
+        """Modify the record before writing to the stream."""
+        record = self.add_object_to_id(record)
+        return record
+
+    def append_times_to_dates(self, record: Dict):
+        """Append times to date fields."""
+        for date_field in self.date_fields:
+            if record.get(date_field):
+                record[date_field] = strftime(strptime_with_tz(record[date_field]))
+
+    def get_url_endpoint(self, parent_obj: Dict = None) -> str:
+        """Get the URL endpoint for the stream."""
+        return self.url_endpoint or f"{self.client.base_url}/{self.path}"
+
+
+class IncrementalStream(BaseStream):
+    """Base Class for Incremental Stream."""
+
+    def get_bookmark(self, state: dict, stream: str, key: Any = None) -> int:
+        """A wrapper for singer.get_bookmark to deal with compatibility for
+        bookmark values or start values."""
+        return get_bookmark(
+            state,
+            stream,
+            key or self.replication_keys[0],
+            self.client.config["start_date"],
+        )
+
+    def write_bookmark(
+        self, state: dict, stream: str, key: Any = None, value: Any = None
+    ) -> Dict:
+        """A wrapper for singer.get_bookmark to deal with compatibility for
+        bookmark values or start values."""
+        if not (key or self.replication_keys):
+            return state
+
+        current_bookmark = get_bookmark(
+            state,
+            stream,
+            key or self.replication_keys[0],
+            self.client.config["start_date"],
+        )
+        value = max(current_bookmark, value)
+        return write_bookmark(state, stream, key or self.replication_keys[0], value)
```

Comment:
This method seems calling itself recursively, may run into infinite loop.

Reply:
the method that its calling is the singer -> write bookmark, if it was calling itself it should be using self.write_bookmark.

this might not be a issue

Reply:
Correct!! @vishalp-dev